### PR TITLE
chore: Refactoring entity types and updating DS action create permission to fix some bugs on EE

### PR DESCRIPTION
**Pull request by ankitakinger**
## Description

Refactoring entity types and updating DS action create permission to fix some bugs on EE

#### PR fixes following issue(s)
Fixes # (issue number)

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

#### How Has This Been Tested?
- [x] Manual
- [ ] JUnit
- [x] Jest
- [x] Cypress

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **Refactor**
  - Streamlined entity type naming conventions across the application for better consistency.
  - Enhanced type definitions for improved code clarity and maintainability.
  - Updated function calls to use object parameters with named properties for better readability.

- **New Features**
  - Introduced a new entity type for module instances, expanding the application's data handling capabilities.

- **Bug Fixes**
  - Corrected improper type assertions to ensure accurate entity recognition and processing.

- **Documentation**
  - Added comments to clarify the non-introduction of certain dependencies in specific components.

- **Style**
  - Adjusted import statements to align with the updated naming conventions.

- **Tests**
  - Updated test cases to reflect changes in entity type references and assertions.

### CODE REVIEW

File: app/client/src/utils/autocomplete/CodemirrorTernService.ts [hunk 1]
```diff
@@ -95,7 +95,7 @@ interface RequestQuery {
 }
 
 export interface DataTreeDefEntityInformation {
-  type: ENTITY_TYPE;
+  type: EntityTypeValue;
   subType: string;
 }
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/utils/autocomplete/CodemirrorTernService.ts]: Existing tests cover this change

File: app/client/src/pages/Editor/DatasourceInfo/DatasourceViewModeSchema.tsx [hunk 2]
```diff
@@ -73,16 +75,19 @@ const DatasourceViewModeSchema = (props: Props) => {
...
+  const canCreateDatasourceActions = hasCreateDSActionPermissionInApp({
+    isEnabled: isFeatureEnabled,
+    dsPermissions: datasourcePermissions,
     pagePermissions,
-  );
+    editorType,
+  });
 
   const applicationId: string = useSelector(getCurrentApplicationId);
   const { pageId: currentPageId } = useParams<ExplorerURLParams>();
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/DatasourceInfo/DatasourceViewModeSchema.tsx]: The typescript added for the params and the permissions are explanatory.

File: app/client/src/pages/Editor/DataSourceEditor/hooks.ts [hunk 3]
```diff
@@ -146,15 +148,18 @@ export const useShowPageGenerationOnHeader = (
...
-  const canCreateDatasourceActions = hasCreateDSActionPermissionInApp(
-    isGACEnabled,
-    datasourcePermissions,
+  const canCreateDatasourceActions = hasCreateDSActionPermissionInApp({
+    isEnabled: isGACEnabled,
+    dsPermissions: datasourcePermissions,
     pagePermissions,
-  );
+    editorType,
+  });
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/DataSourceEditor/hooks.ts]: I have updated all the usages of the function `hasCreateDSActionPermissionInApp`

File: app/client/src/utils/autocomplete/__tests__/TernServer.test.ts [hunk 4]
```diff
@@ -254,11 +254,11 @@ describe("Tern server sorting", () => {
...
   defEntityInformation.set("sameEntity", {
-    type: ENTITY_TYPE_VALUE.WIDGET,
+    type: ENTITY_TYPE.WIDGET,
     subType: "TABLE_WIDGET",
   });
   defEntityInformation.set("sameEntity", {
-    type: ENTITY_TYPE_VALUE.WIDGET,
+    type: ENTITY_TYPE.WIDGET,
     subType: "TABLE_WIDGET_V2",
   });
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/utils/autocomplete/__tests__/TernServer.test.ts]: The tests for the same are covered in integration tests

File: app/client/src/components/editorComponents/GlobalSearch/utils.tsx [hunk 5]
```diff
@@ -88,19 +88,20 @@ export interface SnippetBody {
...
 
 export const getSnippetFilterLabel = (state: AppState, label: string) => {
   return (
     WidgetFactory.widgetConfigMap.get(label as WidgetType)?.widgetName ||
     getPluginByPackageName(state, label)?.name ||
-    filterEntityTypeLabels[label as ENTITY_TYPE] ||
+    filterEntityTypeLabels[label as EntityTypeValue] ||
     label
   );
 };
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/components/editorComponents/GlobalSearch/utils.tsx]: When we are using it as a type we are using `EntityTypeValue` while to use it as a value we are using `ENTITY_TYPE`

File: app/client/src/ce/components/editorComponents/GPT/trigger.tsx [hunk 6]
```diff
@@ -17,7 +17,7 @@ export const getAIContext = ({
   cursorPosition,
   editor,
 }: {
-  entityType?: ENTITY_TYPE;
+  entityType?: EntityTypeValue;
   slashIndex?: number;
   currentLineValue?: string;
   cursorPosition: CodeMirror.Position;
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/components/editorComponents/GPT/trigger.tsx]: It is used in EE

File: app/client/src/entities/DataTree/dataTreeFactory.ts [hunk 7]
```diff
@@ -3,14 +3,14 @@ import { generateDataTreeJSAction } from "@appsmith/entities/DataTree/dataTreeJS
...
 } from "@appsmith/entities/DataTree/types";
 import { generateDataTreeModuleInputs } from "@appsmith/entities/DataTree/utils";
 import type {
   DataTreeSeed,
   AppsmithEntity,
-  ENTITY_TYPE,
+  EntityTypeValue,
 } from "@appsmith/entities/DataTree/types";
 import type {
   unEvalAndConfigTree,
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/entities/DataTree/dataTreeFactory.ts]: Existing tests have been updated and some tests are covered in integration tests

File: app/client/src/ce/reducers/entityReducers/index.ts [hunk 8]
```diff
@@ -31,4 +33,6 @@ export const entityReducerObject = {
   autoHeightLayoutTree: autoHeightLayoutTreeReducer,
   canvasLevels: canvasLevelsReducer,
   layoutElementPositions: layoutElementPositionsReducer,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  moduleInstanceEntities: (state: AppState, action: ReduxAction<any>) => {},
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/reducers/entityReducers/index.ts]: This is a placeholder on CE while the reducer logic is on EE alone.

[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/reducers/entityReducers/index.ts]: No tests for this is required on CE

File: app/client/src/ce/plugins/Linting/utils/getEntityUniqueIdForLogs.ts [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
 
 const entityUniqueIdGetterMap: Record<
   string,
   (entity: DataTreeEntity) => string
 > = {
-  [ENTITY_TYPE_VALUE.JSACTION]: (entity) => {
+  [ENTITY_TYPE.JSACTION]: (entity) => {
     return (entity as JSActionEntity).actionId;
   },
 };
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/plugins/Linting/utils/getEntityUniqueIdForLogs.ts:12]: It exists.

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx [hunk 10]
```diff
@@ -216,6 +237,62 @@ export function EntityProperties() {
...
+              return {
+                propertyName: actionProperty,
+                entityName: jsCollection.config.name,
+                value: value,
+                entityType,
+              };
+            },
+          );
+        }
+      }
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx]: Its a combination of `ENTITY_TYPE.ACTION` and `ENTITY_TYPE.JSACTION`

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx [hunk 11]
```diff
@@ -216,6 +237,62 @@ export function EntityProperties() {
...
+      break;
   }
   return (
     <EntityInfoContainer
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx]: Intentional

File: app/client/src/ce/utils/autocomplete/entityDefGeneratorMap.ts [hunk 12]
```diff
@@ -41,30 +41,30 @@ export type EntityDefGeneratorMap = Record<
...
-      subType: ENTITY_TYPE_VALUE.APPSMITH,
+      type: ENTITY_TYPE.APPSMITH,
+      subType: ENTITY_TYPE.APPSMITH,
     });
   },
-  [ENTITY_TYPE_VALUE.JSACTION]: (props) => {
+  [ENTITY_TYPE.JSACTION]: (props) => {
     const {
       configTree,
       def,
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/utils/autocomplete/entityDefGeneratorMap.ts]: Updated the tests already

File: app/client/src/utils/widgetRenderUtils.tsx [hunk 13]
```diff
@@ -7,7 +7,7 @@ import type {
   WidgetEntityConfig,
 } from "@appsmith/entities/DataTree/types";
 import type { ConfigTree, DataTree } from "entities/DataTree/dataTreeTypes";
-import { ENTITY_TYPE_VALUE } from "entities/DataTree/dataTreeFactory";
+import { ENTITY_TYPE } from "entities/DataTree/dataTreeFactory";
 import { pick } from "lodash";
 import {
   WIDGET_DSL_STRUCTURE_PROPS,
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/utils/widgetRenderUtils.tsx]: Done already

File: app/client/src/ce/utils/BusinessFeatures/permissionPageHelpers.tsx [hunk 14]
```diff
@@ -153,13 +154,19 @@ export const getHasAuditLogsReadPermission = (
...
+}: {
+  dsPermissions?: string[];
+  editorType?: string;
+  isEnabled: boolean;
+  pagePermissions?: string[];
+}) => {
+  return !editorType || editorType === EditorNames.APPLICATION
+    ? getHasCreateDatasourceActionPermission(isEnabled, dsPermissions) &&
+        getHasCreateActionPermission(isEnabled, pagePermissions)
+    : getHasCreateDatasourceActionPermission(isEnabled, dsPermissions);
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/utils/BusinessFeatures/permissionPageHelpers.tsx]: The change done here is intentional

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx [hunk 15]
```diff
@@ -216,6 +237,62 @@ export function EntityProperties() {
...
+                } else {
+                  value = entity.data.body;
+                }
+              }
+              return {
+                propertyName: actionProperty,
+                entityName: entityName,
+                value,
+                entityType,
+              };
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx]: intentional

File: app/client/src/ce/entities/DataTree/types.ts [hunk 16]
```diff
@@ -177,7 +180,7 @@ export interface WidgetEntityConfig
 }
 
 export interface AppsmithEntity extends Omit<AppDataState, "store"> {
-  ENTITY_TYPE: ENTITY_TYPE.APPSMITH;
+  ENTITY_TYPE: typeof ENTITY_TYPE.APPSMITH;
   store: Record<string, unknown>;
   theme: AppTheme["properties"];
 }
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/entities/DataTree/types.ts:186]: It was intentionally done by another developer.

File: app/client/src/components/editorComponents/ActionCreator/helpers.tsx [hunk 17]
```diff
@@ -110,7 +110,7 @@ export function getFieldFromValue(
   const entity = dataTree && dataTree[entityProps.entityName];
 
   if (entity && "ENTITY_TYPE" in entity) {
-    if (isAction(entity)) {
+    if (isAction(entity as DataTreeEntity)) {
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/components/editorComponents/ActionCreator/helpers.tsx:113]: Its intentional change

File: app/client/src/ce/reducers/entityReducers/index.ts [hunk 18]
```diff
@@ -31,4 +33,8 @@ export const entityReducerObject = {
   autoHeightLayoutTree: autoHeightLayoutTreeReducer,
   canvasLevels: canvasLevelsReducer,
   layoutElementPositions: layoutElementPositionsReducer,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  moduleInstanceEntities: (state: AppState, action: ReduxAction<any>) => {
```
Comment (ashit-rath) [inline app/client/src/ce/reducers/entityReducers/index.ts:37]: What is this for?

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts [hunk 19]
```diff
@@ -1048,7 +1051,7 @@ function* executeOnPageLoadJSAction(pageAction: PageAction) {
   }
 
   const jsAction = collection.actions.find(
-    (action) => action.id === pageAction.id,
+    (action: JSAction) => action.id === pageAction.id,
```
Comment (ashit-rath) [inline app/client/src/sagas/ActionExecution/PluginActionSaga.ts:1054]: Doesn't JSAction apply implicitly?

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx [hunk 20]
```diff
@@ -42,6 +43,70 @@ const EntityInfoContainer = styled.div`
...
+        entityName: jsCollection.config.name,
+        value: value,
+        entityType,
+      };
+    });
+  }
+  return entityProperties;
+};
+
+const getActionBindings = (
```
Comment (ashit-rath) [inline app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx:67]: Are these changes just normal refactor or have we added more logic to it?

File: app/client/src/ce/hooks/index.ts [hunk 21]
```diff
@@ -5,7 +5,7 @@ import {
 import { matchPath } from "react-router";
 
 export const EditorNames = {
-  APPLICATION: "appEditor",
+  APPLICATION: "app",
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/hooks/index.ts:8]: to use it directly here:
image

File: app/client/src/ce/reducers/entityReducers/index.ts (hunk 18 above)
Comment (ankitakinger) [reply app/client/src/ce/reducers/entityReducers/index.ts:37]: The type changes (ENTITY_TYPE) were causing issues in common code where we have directly used `state.entities.moduleInstanceEntities`

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts (hunk 19 above)
Comment (ankitakinger) [reply app/client/src/sagas/ActionExecution/PluginActionSaga.ts:1054]: No typescript was giving error, so I added the type

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx (hunk 20 above)
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx:67]: just normal refactor to use it as is for module instance entities binding.

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts (hunk 19 above)
Comment (ashit-rath) [reply app/client/src/sagas/ActionExecution/PluginActionSaga.ts:1054]: What was the error?

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx [hunk 22]
```diff
@@ -129,73 +209,35 @@
...
 
-  const entity: any = widgetEntity || actionEntity || jsActionEntity;
-  let config: any;
+  const entity: any =
+    widgetEntity ||
+    actionEntity ||
+    jsActionEntity ||
+    moduleInstanceQueryEntity ||
+    moduleInstanceJSEntity;
   let entityProperties: any = [];
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx:218]: intentional

File: app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx [hunk 23]
```diff
@@ -42,6 +43,70 @@ const EntityInfoContainer = styled.div`
...
+      return {
+        propertyName: actionProperty,
+        entityName: jsCollection.config.name,
+        value: value,
+        entityType,
+      };
+    });
+  }
+  return entityProperties;
+};
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx:65]: intentional

File: app/client/src/sagas/ActionExecution/PluginActionSaga.ts (hunk 19 above)
Comment (ankitakinger) [reply app/client/src/sagas/ActionExecution/PluginActionSaga.ts:1054]: Parameter implicitly has `any` type

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**